### PR TITLE
Issue #2872 Remove entries from kubeconfig file on cluster deletion

### LINKED ISSUE #2872: Deleting a Minishift instance should cleanup kubeconfg
Currently when the Minishift instance gets deleted its state as kept in MINISHIFT_HOME gets deleted. However, the kubeconfig under _~/.kube_ stays untouched. This over time clutters the configuration. 

I think it makes sense to clean the kubeconfig the moment the Minishift instance gets deleted.

### PR DESCRIPTION
**Pull request by anjannath**
Fixes #2872 

Steps to test the pull request

  1. `minishift start`
  2. `minishift delete`
  3. Check `~/.kube/config` before and after deletion to verify that the entries corresponding to the current cluster are removed.

### CODE REVIEW

File: cmd/minishift/cmd/delete.go [hunk 1]
```diff
@@ -111,6 +121,59 @@ func clearCache() {
...
+		return err
+	}
+	cleanKubeConfig, err := removeEntriesForCluster(hostIP, kubeConfig)
+	if err != nil {
+		return err
+	}
+	return clientcmd.WriteToFile(*cleanKubeConfig, kubeConfigPath)
+}
+
+func removeEntriesForCluster(hostIP string, kubeConfig *clientcmdapi.Config) (*clientcmdapi.Config, error) {
```
Comment (anjannath) [inline cmd/minishift/cmd/delete.go]: I am not able to write the unit test for this function. any ideas @praveenkumar

Comment (anjannath) [reply cmd/minishift/cmd/delete.go]: @praveenkumar I tried that approach, but we'd need two different kubeconfig files, one _dirty_ and one _clean_ to compare that the o/p of the function is expected. Now the structs inside the config struct has this `LocationOforigin` field which contains the fullpath of the kubeconfig file, which will be different. So i could not make this work.

Comment (anjannath) [reply cmd/minishift/cmd/delete.go]: @praveenkumar Went with renaming the files, it works now!

File: cmd/minishift/cmd/delete_test.go [hunk 2]
```diff
@@ -71,6 +73,35 @@ func Test_clear_cache_forced(t *testing.T) {
...
+
+	dirtyKubeConfigPath := filepath.Join("..", "..", "..", "test", "testdata", "kubeconfig_dirty")
+	cleanKubeConfigPath := filepath.Join("..", "..", "..", "test", "testdata", "kubeconfig_clean")
+
+	dirtyKubeConfig, err := clientcmd.LoadFromFile(kubeConfigPath)
+	assert.NoError(t, err, "Error loading dirty kubeconfig file")
+
+	os.Rename(kubeConfigPath, dirtyKubeConfigPath)
+	os.Rename(cleanKubeConfigPath, kubeConfigPath)
+
```
Comment (anjannath) [inline cmd/minishift/cmd/delete_test.go]: Moving the files, to keep the `LocationOfOrigin` field same in the config structs.

File: cmd/minishift/cmd/delete_test.go [hunk 3]
```diff
@@ -71,6 +73,35 @@ func Test_clear_cache_forced(t *testing.T) {
...
+	cleanKubeConfig, err := clientcmd.LoadFromFile(kubeConfigPath)
+	assert.NoError(t, err, "Error loading clean kubeconfig file")
+
+	os.Rename(kubeConfigPath, cleanKubeConfigPath)
+	os.Rename(dirtyKubeConfigPath, kubeConfigPath)
+
```
Comment (anjannath) [inline cmd/minishift/cmd/delete_test.go]: Restoring again after..

**Comment by agajdosi** [comment]
retest this please

**Comment by agajdosi** [comment]
Tried out on Fedora and after the deletion of Minishift, the config stays the same as when cluster was running. Seems like nothing is removed from the config.

**Comment by anjannath** [comment]
@agajdosi It'll only remove the entries corresponding to the active profiles cluster, if you have lot of entries in the kubeconfig it is hard to track that..

File: cmd/minishift/cmd/delete.go [hunk 4]
```diff
@@ -82,6 +86,12 @@ func runDelete(cmd *cobra.Command, args []string) {
 			atexit.ExitWithMessage(1, err.Error())
 		}
 	}
+	// Remove entries from global kube config
+	clusterIP, _ := cluster.GetHostIP(api)
+	err = clearKubeConfig(clusterIP)
```
Comment (gbraad) [inline cmd/minishift/cmd/delete.go]: not a good name... it does not clear.. it removes entries

File: cmd/minishift/cmd/delete.go [hunk 5]
```diff
@@ -111,6 +121,59 @@ func clearCache() {
...
+func clearKubeConfig(clusterIP string) error {
+	kubeConfigPath, err := oc.GetGlobalKubeConfigPath()
+	if err != nil {
+		return err
+	}
+	kubeConfig, err := clientcmd.LoadFromFile(kubeConfigPath)
+	if err != nil {
+		return err
+	}
+	cleanKubeConfig, err := removeEntriesForCluster(clusterIP, kubeConfig)
```
Comment (gbraad) [inline cmd/minishift/cmd/delete.go]: and here you call it `cleanKubeConfig`

**Comment by praveenkumar** [comment]
merged after fixing the function typo.

**Pull request closed without merging**